### ISSUE TITLE
Support Kotlin PanacheCompanion / PancheEntity for reactive pattern

### ISSUE THREAD

**Issue by ttsatsanis** [quarkusio/quarkus#16785]
## Description

Currently `PanacheCompanion` does not support `io.quarkus.hibernate.reactive.panache.PanacheEntityBase`.
It would be very helpful to create interfaces for `PanacheCompanion` and `PanacheEntityBase` / `PanacheEntity` on a reactive stack of `io.quarkus.hibernate.orm.panache.kotlin`

Reactive pattern on Kotlin can only be implemented through `PanacheRepository` or by introducing static methods which return `Uni` or `Multi` (which is a bad practice, since they do not depend on the non-blocking `Mutiny.Session`)

**Comment by evanchooly** [comment]
There is already this module: `hibernate-orm-panache-kotlin`

**Comment by evanchooly** [comment]
In my head, I thought I'd already done this but clearly I was thinking of the reactive mongodb module.

**Comment by ttsatsanis** [comment]
@evanchooly yes, I was referring to implementations apart from Mongo which already has a reactive stack with Panache on Kotlin (I currently use Hibernate with Postgres)

**Comment by MartinX3** [comment]
Yes, it would be very useful to use kotlin with panache hibernate reactive.

**Comment by MartinX3** [comment]
> Is there a workaround or example on how to get by without it at the moment?

In my case just using
- "io.quarkus:quarkus-jdbc-postgresql"
- "io.quarkus:quarkus-hibernate-orm-panache-kotlin"
- PanacheRepositoryBase<YourEntityClass, UUID> 
- kotlin data classes as Entities

So having a not reactive database access layer under the service layer while having resteasy with jackson as reactive components in the api layer.

**Comment by nenros** [comment]
you can still use java reactive version, it needs bit more code, but it should work

**Comment by MartinX3** [comment]
> you can still use java reactive version, it needs bit more code, but it should work

yes, especially the null handling.

**Comment by nenros** [comment]
https://github.com/quarkusio/quarkus/pull/21321 @evanchooly I tried to do something in that topic but i'm not good enough, lot of tests is not working, and I still don't know how I should handle session from original Panache class. If you could take look, and say what i should take look.

**Comment by kenyee** [comment]
@MartinX3 : can you post a small snippet of how your resource calls into your service?
I'm using a suspend fun resource but getting this error: "You have attempted to perform a blocking operation on a IO thread" which isn't intuitive.  It's a suspend method, so it should be able to block, though ideally would run using a reactive Panache call but it seems like "io.quarkus:quarkus-hibernate-orm-panache-kotlin" is non-reactive.

**Comment by MartinX3** [comment]
@kenyee I use the Java version of the reactive stuff and add the needed workarounds intelliJ told me to use it.
Also I use reactive in every layer and Mutiny for the data transfer between the layers.

Also I don't use any await, just mutiny stuff.

If you don't use unis or multis, you need to put @Blocked on every rest API class or whatever interacts with your application.

**Comment by kenyee** [comment]
Thanks...was hoping to use kotlin coroutines instead because it looked like resteasy reactive supported using suspend fun methods 🤷 
I'll try using the non-kotlin panache-reactive instead.

**Comment by kenyee** [comment]
and any idea why we need a -kotlin version of quarkus-hibernate-orm-panache?
Seems like Kotlin should just work in backwards compatibility mode...

**Comment by MartinX3** [comment]
You can always use java stuff, but kotlin stuff is optimized for the kotlin language.
As ex ample you get kotlin nullables.

**Comment by loicmathieu** [comment]
@kenyee the Kotlin version provide an idiomatic Kotlin API but you can use the Java version.

For coroutines you need to use  reactive Resteasy, however to help you we will need some reproducer project to understand what you're trying to do.

**Comment by kenyee** [comment]
Attached a the small sample here:
https://github.com/quarkusio/quarkus/discussions/23122

In ProjectResource.kt, just change the "fun" to "suspend fun".  My expectation was that it would run as a coroutine on the computation threads and then I could do a withContext(Dispatchers.io) to run the non-reactive hibernate code on the IO thread (not on a Vertx I/O thread, but a Kotlin I/O thread).

Ideally, I'd like to have used a version of quarkus-hibernate-orm-panache that could work w/ reactive hibernate.
For the idiomatic Kotlin API for panache, are there diffs vs. the non-java one?  Curious what was changed.  You can mark Java methods with @Nullable and @NotNull and IntelliJ will make them work properly w/ Kotlin so you don't need specific classes.  Only other expectations would be support for coroutines and maybe a few extension methods to make it nicer, but those aren't really related to Panache AFAIK...

I also noticed this warning:
```
10:20:20,619 WARN  [io.qua.hib.orm.run.pro.ProxyDefinitions] Could not generate an enhanced proxy for entity 'org.twitter.teamcodehealth.repository.Project' (class='org.twitter.teamcodehealth.repository.Project') as it's final. Your application might perform better if we're allowed to extend it.
Daemon will be stopped at the end of the build after running out of JVM memory
```
The allopen plugin should probably be configured to open anything w/ the @Entity annotation automatically.

**Comment by kenyee** [comment]
@MartinX3 oops...thanks...forgot to add that.  It's a new project so I haven't hooked in the DB all the way yet...

**Comment by kenyee** [comment]
FWIW, I looked over the code in https://github.com/quarkusio/quarkus/tree/main/extensions/panache/hibernate-orm-panache-kotlin
and don't see anything special about it... you can probably get rid of it and annotate the Java version of these with @Nullable and @NotNull and it'll work fine w/ Kotlin.  There are hard nullability kotlin declarations with Kotlin code, but this isn't strictly needed because the compiler will use the @Nullable/NotNull Java annotations to project the same APIs in interop mode.

The reactive pachane support might want to have a Kotlin module that maps Mutiny APIs to Kotlin suspend functions though..but that would be on top of the Mutiny support.

**Comment by rohit-s8** [comment]
Are there any plans to support this coming soon?

**Comment by evanchooly** [comment]
fixed by https://github.com/quarkusio/quarkus/pull/25015
